fix position of tooltip in screen grabber

antis81 opened the pull request:
The tooltip is shown centred at top of the screen which currently contains the mouse cursor.

antis81 commented:
@PKEv Bet you will like this… :)

PKEv commented:
looks like this
![default](https://cloud.githubusercontent.com/assets/6053197/10188893/9bbd6406-676b-11e5-84b3-0902f2988227.png)

added:
screenRect - contain only size of 1 monitor

antis81 commented:
At least it's on the right screen (or is it?). :smile:

Anyway I'll put some qDebug's inside the next commit to investigate, where the x-position is messed up on Windows. My guess is, that `QScreen::geometry()` returns different values on Windows (<=7,8 ???). But let's see… @PKEv Can you please fetch and write the output values on Windows?

Mine are (for right==secondary screen):

```
[13:14:15.902] ../../qtox/src/widget/tool/screenshotgrabber.cpp:202 : Debug: screenRect == QRect(1279,0 1600x900)
[13:14:15.902] ../../qtox/src/widget/tool/screenshotgrabber.cpp:203 : Debug: screenRect.center() == QPoint(2078,449)
[13:14:15.902] ../../qtox/src/widget/tool/screenshotgrabber.cpp:206 : Debug: ttRect == QRectF(0,0 967.312x23)
[13:14:15.902] ../../qtox/src/widget/tool/screenshotgrabber.cpp:208 : Debug: x == 1594
```

Btw.: I checked this on Linux (X11) in two multi-monitor setups (with primary screen left/right). It simply works in any combination of screens…

PKEv commented:
when 1 screen:

```
[15:27:06.267] ../qTox/src/widget/tool/screenshotgrabber.cpp:202 : Debug: screenRect == QRect(0,0 1920x1080)
[15:27:06.267] ../qTox/src/widget/tool/screenshotgrabber.cpp:203 : Debug: screenRect.center() == QPoint(959,539)
[15:27:06.267] ../qTox/src/widget/tool/screenshotgrabber.cpp:206 : Debug: ttRect == QRectF(0,0 655x21)
[15:27:06.267] ../qTox/src/widget/tool/screenshotgrabber.cpp:208 : Debug: x == 631
[15:27:06.297] ../qTox/src/widget/tool/screenshotgrabber.cpp:202 : Debug: screenRect == QRect(0,0 1920x1080)
[15:27:06.297] ../qTox/src/widget/tool/screenshotgrabber.cpp:203 : Debug: screenRect.center() == QPoint(959,539)
[15:27:06.297] ../qTox/src/widget/tool/screenshotgrabber.cpp:206 : Debug: ttRect == QRectF(0,0 655x21)
[15:27:06.297] ../qTox/src/widget/tool/screenshotgrabber.cpp:208 : Debug: x == 631
```

when 2 screen, the main on the right side:

```
[15:28:26.826] ../qTox/src/widget/tool/screenshotgrabber.cpp:202 : Debug: screenRect == QRect(0,0 1920x1080)
[15:28:26.826] ../qTox/src/widget/tool/screenshotgrabber.cpp:203 : Debug: screenRect.center() == QPoint(959,539)
[15:28:26.826] ../qTox/src/widget/tool/screenshotgrabber.cpp:206 : Debug: ttRect == QRectF(0,0 655x21)
[15:28:26.826] ../qTox/src/widget/tool/screenshotgrabber.cpp:208 : Debug: x == 631
[15:28:26.879] ../qTox/src/widget/tool/screenshotgrabber.cpp:202 : Debug: screenRect == QRect(0,0 1920x1080)
[15:28:26.879] ../qTox/src/widget/tool/screenshotgrabber.cpp:203 : Debug: screenRect.center() == QPoint(959,539)
[15:28:26.879] ../qTox/src/widget/tool/screenshotgrabber.cpp:206 : Debug: ttRect == QRectF(0,0 655x21)
[15:28:26.879] ../qTox/src/widget/tool/screenshotgrabber.cpp:208 : Debug: x == 631
```

when 2 screen, the main on the left side:

```
[15:28:54.825] ../qTox/src/widget/tool/screenshotgrabber.cpp:202 : Debug: screenRect == QRect(0,0 1920x1080)
[15:28:54.825] ../qTox/src/widget/tool/screenshotgrabber.cpp:203 : Debug: screenRect.center() == QPoint(959,539)
[15:28:54.825] ../qTox/src/widget/tool/screenshotgrabber.cpp:206 : Debug: ttRect == QRectF(0,0 655x21)
[15:28:54.825] ../qTox/src/widget/tool/screenshotgrabber.cpp:208 : Debug: x == 631
[15:28:54.875] ../qTox/src/widget/tool/screenshotgrabber.cpp:202 : Debug: screenRect == QRect(0,0 1920x1080)
[15:28:54.875] ../qTox/src/widget/tool/screenshotgrabber.cpp:203 : Debug: screenRect.center() == QPoint(959,539)
[15:28:54.875] ../qTox/src/widget/tool/screenshotgrabber.cpp:206 : Debug: ttRect == QRectF(0,0 655x21)
[15:28:54.875] ../qTox/src/widget/tool/screenshotgrabber.cpp:208 : Debug: x == 631
```

As a result:
Calculated always only the size of the main display.
(only virtualGeometry() retern full size)

edded:
the main display always first in this cycle:

```
    foreach(QScreen* s, QApplication::screens()) {
        QRect g = s->geometry();
        if (g.contains(mousePos)) {
            screenRect = g;
            break;
        }
    }
```

antis81 commented:
@PKEv I'm a bit puzzled about this. Can you please copy the following code and put it in front of the foreach loop? It just outputs the two screen geometries. Of course you can extend it and output e.g. mouse cursor position to get a more complete picture of the situation.

``` c++
    // >>>>> TEST CODE
    const QPoint mousePos = QCursor::pos();
    QList<QScreen*> screens = QApplication::screens();
    for(int i=0; i < screens.count(); i++) {
        qDebug() << "Screen" << i << "has geometry" << screens[i]->geometry();
        // these should be the global positions (within the whole virtual desktop area)
    }
    return;
    // <<<<< TEST CODE

    QRect screenRect;
    foreach(QScreen* s, QApplication::screens()) {
        QRect g = s->geometry();
        if (g.contains(mousePos)) {
            screenRect = g;
            break;
        }
    }
...
...
```

**EDIT:** fixed bugs from blind writing :innocent:

PKEv commented:
```
[16:49:26.479] ../qTox/src/widget/tool/screenshotgrabber.cpp:200 : Debug: Screen 0 has geometry QRect(0,0 1920x1080)
[16:49:26.479] ../qTox/src/widget/tool/screenshotgrabber.cpp:200 : Debug: Screen 1 has geometry QRect(-1280,0 1280x800)
```

antis81 commented:
Thanks! I also did that. Found, that X11 (Qt on Linux?) transforms the screen geometries always (!) starting from P(0,0) - never negative.

Values for "primary screen on right side":

```
[15:40:17.689] ../../qtox/src/widget/tool/screenshotgrabber.cpp:195 : Debug: Screen 0 has geometry QRect(1599,0 1280x800)
[15:40:17.689] ../../qtox/src/widget/tool/screenshotgrabber.cpp:195 : Debug: Screen 1 has geometry QRect(0,0 1600x900)
[15:40:17.689] ../../qtox/src/widget/tool/screenshotgrabber.cpp:196 : Debug: Mouse pos is QPoint(399,217)
```

See the difference? On Linux (X11 actually), screen coordinates always match the virtual desktop coordinates - that's why we don't have problems here… :)

Hm… any idea, how to transform them correctly independent of platform?

antis81 commented:
@PKEv Think I found solution. Can you check with `QScreen::virtualSiblings()`?

`````` c++
void ScreenshotGrabber::adjustTooltipPosition()
{
    const QPoint mousePos = QCursor::pos();

    // according to Qt doc, virtualSiblings should work…
    QList<QScreen*> screens = QGuiApplication::primaryScreen()->virtualSiblings();

    // >>>>> TEST CODE
    for(int i=0; i < screens.count(); i++) {
        qDebug() << "Screen" << i << "has geometry" << screens[i]->geometry();
        // these should be the global positions (within the whole virtual desktop area)
    }
    qDebug() << "Mouse pos is" << mousePos;
//    return;
    // <<<<< TEST CODE

    QRect screenRect;
    foreach(QScreen* s, screens) {
        QRect g = s->geometry();
        if (g.contains(mousePos)) {
            screenRect = g;
            break;
        }
    }
    Q_ASSERT(screenRect.isValid());

    qDebug() << "screenRect ==" << screenRect;
    qDebug() << "screenRect.center() ==" << screenRect.center();

    const QRectF ttRect = this->helperToolbox->childrenBoundingRect();
    qDebug() << "ttRect ==" << ttRect;
    int x = screenRect.center().x() - (ttRect.width() / 2);
    qDebug() << "x ==" << x;
    helperToolbox->setX(x);
    helperToolbox->setY(screenRect.y());
}```
``````

PKEv commented:
```
[18:02:58.327] ../qTox/src/widget/tool/screenshotgrabber.cpp:201 : Debug: Screen 0 has geometry QRect(0,0 1920x1080)
[18:02:58.327] ../qTox/src/widget/tool/screenshotgrabber.cpp:201 : Debug: Screen 1 has geometry QRect(-1280,0 1280x800)
[18:02:58.327] ../qTox/src/widget/tool/screenshotgrabber.cpp:204 : Debug: Mouse pos is QPoint(930,542)
```

edded:
Values for "primary screen on right side", too

PKEv commented:
Do you want to place text in the middle of the screen? - use virtualGeometry
get the full picture width and use it as the coordinates

antis81 commented:
Sorry, I assumed it was clear what I meant. The text shall appear horizontally centred at top of the "current" screen (the one containing the mouse cursor). This is very likely the screen, where the human eyes are focused on. It automagically updates, when screen borders are crossed.

Can you (force) fetch please and check the next commit? And would you confirm, that `desk->mapFromGlobal` works for mapping the global (positive) mouse position (should be negative on your default setup in Windows with primary screen on the right). In X11, local/global positions are the same.

PKEv commented:
```
[00:04:57.847] ../qTox/src/widget/tool/screenshotgrabber.cpp:194 : Debug: Global mouse pos. = QPoint(1271,700)
[00:04:57.847] ../qTox/src/widget/tool/screenshotgrabber.cpp:195 : Debug: Local mouse pos. = QPoint(1271,700)
[00:04:57.848] ../qTox/src/widget/tool/screenshotgrabber.cpp:196 : Debug: Screen rect = QRect(0,0 1366x768)
[00:04:57.848] ../qTox/src/widget/tool/screenshotgrabber.cpp:200 : Debug: x == 354
[00:04:57.899] ../qTox/src/widget/tool/screenshotgrabber.cpp:194 : Debug: Global mouse pos. = QPoint(1271,700)
[00:04:57.905] ../qTox/src/widget/tool/screenshotgrabber.cpp:195 : Debug: Local mouse pos. = QPoint(1271,700)
[00:04:57.905] ../qTox/src/widget/tool/screenshotgrabber.cpp:196 : Debug: Screen rect = QRect(0,0 1366x768)
[00:04:57.905] ../qTox/src/widget/tool/screenshotgrabber.cpp:200 : Debug: x == 354
```

![default](https://cloud.githubusercontent.com/assets/6053197/10206636/9a229e3a-67d0-11e5-8c9f-f9b2def2ad8d.png)

antis81 commented:
Ok, now at least `screenRect` is positive (which goes contrary to QScreen::geometry()). So:

Horizontal centre of (left?) screen is 
1366 / 2 = 683

Width of tool tip is 967 (after implicit cast to int!)
967 / 2 = 483

and finally:
683 - 483 = 200

How the heck does windows get to 354?

Need to sleep for now… Maybe you can fix the rest on Windows??

PKEv commented:
Test this please, it work for me

```
void ScreenshotGrabber::adjustTooltipPosition()
{
    QRect recGL = QGuiApplication::primaryScreen()->virtualGeometry();
    QRect rec = qApp->desktop()->availableGeometry(QCursor::pos());
    const QRectF ttRect = this->helperToolbox->childrenBoundingRect();
    int x = abs(recGL.x())+ rec.x() + rec.width()/2 - (ttRect.width() / 2);
    int y = abs(recGL.y())+ rec.y();
    helperToolbox->setX(x);
    helperToolbox->setY(y);
}
```

antis81 commented:
@PKEv Looks ok. Cleaned it up a little and pushed. IMO `availableGeometry()` should be replaced with `screenGeometry()`. Can you check please?

antis81 commented on src/widget/tool/screenshotgrabber.cpp:
```diff
@@ -181,14 +181,20 @@ void ScreenshotGrabber::chooseHelperTooltipText(QRect rect)
-    helperToolbox->setX(screenRect.x() + (screenRect.width() - size.width() + size.x()) / 2);
-    helperToolbox->setY(screenRect.y());
+    QRect recGL = QGuiApplication::primaryScreen()->virtualGeometry();
+    QRect rec = qApp->desktop()->availableGeometry(QCursor::pos());
```
I mean this line. Replace with `QRect rec = qApp->desktop()->screenGeometry(QCursor::pos());` and we should get perfect results.

PKEv commented:
yes, all good. And it's even better

antis81 committed:
fix position of tooltip in screen grabber

The tooltip is shown centred at top of the screen which currently contains the mouse cursor.
